polaris-core/src/main/java/org/apache/polaris/core/persistence/secrets/PrincipalSecretsGenerator.java

```diff
@@ -38,48 +37,48 @@
+    this.realmName = null;
+  }
+
+  public PrincipalSecretsGenerator(@Nullable String realmName) {
+    this.realmName = realmName;
+  }
 
   /**
    * Produces a new {@link PolarisPrincipalSecrets} object for the given principal ID. The returned
-   * secrets may or may not be random, depending on context. In bootstrapping contexts, the returned
-   * secrets can be predefined. After bootstrapping, the returned secrets can be expected to be
-   * cryptographically random.
+   * secrets may or may not be random, depending on context. The returned secrets can be predefined.
    *
    * @param principalName the name of the related principal. This parameter is a hint for
    *     pre-defined secrets lookup during bootstrapping it is not included in the returned data.
    * @param principalId the ID of the related principal. This ID is part of the returned data.
    * @return a new {@link PolarisPrincipalSecrets} instance for the specified principal.
    */
-  PolarisPrincipalSecrets produceSecrets(@NotNull String principalName, long principalId);
-
-  static PrincipalSecretsGenerator bootstrap(String realmName) {
-    return bootstrap(realmName, System.getenv()::get);
-  }
+  public abstract PolarisPrincipalSecrets produceSecrets(
+      @NotNull String principalName, long principalId);
 
-  static PrincipalSecretsGenerator bootstrap(String realmName, Function<String, String> config) {
-    return (principalName, principalId) -> {
-      String propId = String.format("POLARIS_BOOTSTRAP_%s_%s_CLIENT_ID", realmName, principalName);
-      String propSecret =
-          String.format("POLARIS_BOOTSTRAP_%s_%s_CLIENT_SECRET", realmName, principalName);
+  /**
+   * @param principalName the name of the related principal. This parameter is a hint for
+   *     pre-defined secrets lookup during bootstrapping it is not included in the returned data.
+   * @return true if the secrets generated by this {@link PrincipalSecretsGenerator} are
+   *     Polaris-generated as opposed to being provided by the user or another system.
+   */
+  public abstract boolean systemGeneratedSecrets(@NotNull String principalName);
 
-      String clientId = config.apply(propId.toUpperCase(Locale.ROOT));
-      String secret = config.apply(propSecret.toUpperCase(Locale.ROOT));
-      // use config values at most once (do not interfere with secret rotation)
-      if (clientId != null && secret != null) {
-        return new PolarisPrincipalSecrets(principalId, clientId, secret, secret);
-      } else {
-        return RANDOM_SECRETS.produceSecrets(principalName, principalId);
-      }
-    };
+  /**
+   * Build a PrincipalSecretsGenerator for bootstrapping
+   *
+   * @param realmName the name of the realm
+   * @return A {@link PrincipalSecretsGenerator} that can generate secrets through `produceSecrets`
+   */
+  public static PrincipalSecretsGenerator bootstrap(String realmName) {
+    return new DefaultPrincipalSecretsGenerator(realmName);
```

Comment:
nit: maybe rename `DefaultPrincipalSecretsGenerator` -> `BootstrapPrincipalSecretsGenerator`?.. it is not actually default in `LocalPolarisMetaStoreManagerFactory`

Reply:
Good point. It's just a wrapper, so I was not sure what to call it. Since it's only used during bootstrap, let me change to `BootstrapPrincipalSecretsGenerator` for now